File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {

        ListNode mid = getMid(head);

        ListNode secondHead = reverse(mid.next);

        mid.next = null;

        while(head != null && secondHead != null)
        {
            ListNode temp = head.next;
            head.next = secondHead;
            head = temp;

            ListNode temp1 = secondHead.next;
            secondHead.next = head;
            secondHead = temp1;
        }
        
    }

    private static ListNode getMid(ListNode head)
    {
        ListNode slow = head;
        ListNode fast = head;

        while(fast!= null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }


    private static ListNode reverse(ListNode head)
    {
        ListNode prev = null;
        ListNode current = head;

        while(current != null)
        {
            ListNode next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }

        return prev;
    }
}

File: 0143-reorder-list/0143-reorder-list.js


var reorderList = function(head) {



  if (head == null || head.next == null) return head;

  //  find the mid

  let mid = (head) => {
    let slow = head;
    let fast = head;
    while (fast != null && fast.next != null) {
      slow = slow.next;
      fast = fast.next.next;
    }

    return slow;
  }


  //reverse from the middle

  let reverse = (head) => {
    let prev = null;
    let current = head;
    while (current != null) {
      let next = current.next;
      current.next = prev;
      prev = current;
      current = next;
    }
    return prev;
  }

  let res = (head, secondHead) => {
    let firstHead = head;
    while (firstHead != null && secondHead != null) {
      let f = firstHead.next;
      firstHead.next = secondHead;
      firstHead = f;
      let s = secondHead.next;
      secondHead.next = firstHead;
      secondHead = s;
    }


    return head;
  }

  let middle = mid(head);
  let sHead = reverse(middle.next);
  middle.next = null;
  let ans = res(head, sHead);

  return ans;
};

